src/_P090_MitsubishiHeatPump .ino

```diff
@@ -0,0 +1,698 @@
+	  _bitrate = (_bitrate == 2400 ? 9600 : 2400);
+	  connect(false);
+  }
+}
+
+bool P090_data_struct::read(String& result, bool force) {
+  if (_isUpdatingSettings) {
+    return false;
+  }
+
+  const heatpumpSettings lastSettings = _currentSettings;
+  const heatpumpStatus lastStatus = _currentStatus;
+
+  sync(PACKET_TYPE_DEFAULT);
+
+  if (!_isConnected) {
+    return false;
+  }
+
+  if (force || _currentSettings != lastSettings || _currentStatus.roomTemperature != lastStatus.roomTemperature) {
+    result.reserve(150);
+
+    result = F("{\"roomTemperature\":");
+    result += toString(_currentStatus.roomTemperature, 1);
+    result += F(",\"wideVane\":\"");
+    result += _currentSettings.wideVane;
+    result += F("\",\"power\":\"");
+    result += _currentSettings.power;
+    result += F("\",\"mode\":\"");
+    result += _currentSettings.mode;
+    result += F("\",\"fan\":\"");
+    result += _currentSettings.fan;
+    result += F("\",\"vane\":\"");
+    result += _currentSettings.vane;
+    result += F("\",\"iSee\":");
+    result += boolToString(_currentSettings.iSee);
+    result += F(",\"temperature\":");
+    result += toString(_currentSettings.temperature, 1) + '}';
+
+    return true;
+  }
+
+  return false;
+}
+
+void P090_data_struct::sync(byte packetType) {
+  if((!_isConnected) || (millis() - _lastRecv > (PACKET_SENT_INTERVAL_MS * 10))) {
+    connect(true);
+  }
+  else if(canRead()) {
+    readPacket();
+  }
+  else if(canSend(true)) {
+    byte packet[PACKET_LEN] = {};
+    createInfoPacket(packet, packetType);
+    writePacket(packet, PACKET_LEN);
+  }
+}
+
+bool P090_data_struct::canRead() const {
+  return (_waitForRead && (millis() - PACKET_SENT_INTERVAL_MS) > _lastSend);
+}
+
+bool P090_data_struct::canSend(bool isInfo) const {
+  return (millis() - (isInfo ? PACKET_INFO_INTERVAL_MS : PACKET_SENT_INTERVAL_MS)) > _lastSend;
+}
+
+void P090_data_struct::writePacket(const byte *packet, int length) {
+  for (int i = 0; i < length; i++) {
+     _serial.write(packet[i]);
+  }
+
+#ifdef PLUGIN_090_DEBUG
+  if (loglevelActiveFor(LOG_LEVEL_DEBUG_MORE)) {
+    addLog(LOG_LEVEL_DEBUG_MORE, dumpOutgoingPacket(packet, length));
+  }
+#endif
+
+  _waitForRead = true;
+  _lastSend = millis();
+}
+
+int P090_data_struct::readPacket() {
+  byte header[INFOHEADER_LEN] = {};
+  byte data[PACKET_LEN] = {};
+  bool foundStart = false;
+  int dataSum = 0;
+  byte checksum = 0;
+  byte dataLength = 0;
+
+  _waitForRead = false;
+
+  if(_serial.available() > 0) {
+    // read until we get start byte 0xfc
+    while(_serial.available() > 0 && !foundStart) {
+      header[0] = _serial.read();
+      if(header[0] == HEADER[0]) {
+        foundStart = true;
+        delay(100); // found that this delay increases accuracy when reading, might not be needed though
+      }
+    }
+
+    if(!foundStart) {
+      return RCVD_PKT_FAIL;
+    }
+
+    //read header
+    for(int i=1;i<5;i++) {
+      header[i] = _serial.read();
```

Comment:
You don't know if data is available.
It is best to have a call from PLUGIN_TEN_PER_SECOND to read whatever is present in the serial buffer and feed that per character to a parse function.
So then the calls are really short and it is just feeding the parse function with one or more bytes per run.
The return of such a function should then be true when a valid and complete packet is ready, or false otherwise.
This could then (re)schedule the PLUGIN_READ call if a new value is present.
See plugin for BME280 or GPS for an example of this.

This way you impose minimal load on the ESP node and the chance of missing a packet is also minimal.